File: python/2020/day1.py
from aoc_util import example

M = 2020


def part1(lines):
    """
    >>> part1(example('1'))
    514579
    """
    d = {int(s) for s in lines}
    for x in d:
        y = M - x
        if y in d:
            return x * y


def part2(lines):
    """
    >>> part2(example('1'))
    241861950
    """
    d = [int(s) for s in lines]
    p = {}
    for i in range(len(d)):
        for j in range(i+1, len(d)):
            key = d[i] + d[j]
            if key < M:
                p[key] = d[i] * d[j]
    for x in d:
        y = M - x
        if y in p:
            return x * p[y]

File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2019/Day1.kt
package io.github.ocirne.aoc.year2019

import io.github.ocirne.aoc.AocChallenge

class Day1(val lines: List<String>) : AocChallenge(2019, 1) {

    fun getFuel(mass: Int): Int {
        return mass / 3 - 2
    }

    override fun part1(): Int {
        return lines.sumOf { getFuel(it.toInt()) }
    }

    fun getTotalFuel(initialMass: Int) = sequence {
        var mass = initialMass
        while (true) {
            mass = getFuel(mass)
            if (mass <= 0) {
                break
            }
            yield(mass)
        }
    }.sum()

    override fun part2(): Int {
        return lines.sumOf { getTotalFuel(it.toInt()) }
    }
}
